src/core/server/saved_objects/export/apply_export_transforms.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SavedObject } from '../../../types';
+import { KibanaRequest } from '../../http';
+import { SavedObjectsExportError } from './errors';
+import { SavedObjectsExportTransform, SavedObjectsExportTransformContext } from './types';
+import { getObjKey, SavedObjectComparator } from './utils';
+
+interface ApplyExportTransformsOptions {
+  objects: SavedObject[];
+  request: KibanaRequest;
+  transforms: Record<string, SavedObjectsExportTransform>;
+  sortFunction?: SavedObjectComparator;
+}
+
+export const applyExportTransforms = async ({
+  objects,
+  request,
+  transforms,
+  sortFunction,
+}: ApplyExportTransformsOptions): Promise<SavedObject[]> => {
+  const context = createContext(request);
+  const byType = splitByType(objects);
+
+  let finalObjects: SavedObject[] = [];
+  for (const [type, typeObjs] of Object.entries(byType)) {
+    const typeTransformFn = transforms[type];
+    if (typeTransformFn) {
+      finalObjects = [
+        ...finalObjects,
+        ...(await applyTransform(typeObjs, typeTransformFn, context)),
+      ];
+    } else {
+      finalObjects = [...finalObjects, ...typeObjs];
+    }
+  }
+
+  if (sortFunction) {
+    finalObjects.sort(sortFunction);
+  }
+
+  return finalObjects;
+};
+
+const applyTransform = async (
+  objs: SavedObject[],
+  transformFn: SavedObjectsExportTransform,
+  context: SavedObjectsExportTransformContext
+) => {
+  const objKeys = objs.map(getObjKey);
+  let transformedObjects: SavedObject[];
+  try {
+    transformedObjects = await transformFn(context, objs);
+  } catch (e) {
+    throw SavedObjectsExportError.objectTransformError(objs, e);
```

Comment:
The `SavedObjectExporter` API is public (as the `spaces` plugin is using it for copy to space). That would be introducing monade-based API to core. As I already explained, I don't really mind, but we should probably have a consensus on the `Either` (or more globally FP) pattern/lib we should be using in our public APIs

Reply:
We decided to rather start with implementing `Either` in places we think it could have a high impact and continue to throw errors until we've got more experience with this approach and decide how we will expose it from the API.